File: 2019/day1/day1.js
const fs = require('fs');
const path = require('path');

function calculateFuel(mass) {
  return Math.floor(mass / 3) - 2;
}

function calculateTotalFuel(mass) {
  let fuelNeeded = 0;
  let fuel = mass;

  while (true) {
    fuel = calculateFuel(fuel);
    if (fuel <= 0) {
      break;
    }
    fuelNeeded += fuel;
  }

  return fuelNeeded;
}

const part1 = (input) => {
  const masses = input
    .trim()
    .split('\n')
    .map(s => parseInt(s, 10));

  return masses
    .map(mass => calculateFuel(mass))
    .reduce((fuel, totalFuel) => totalFuel += fuel, 0);
}

const part2 = (input) => {
  const masses = input
    .trim()
    .split('\n')
    .map(s => parseInt(s, 10));

  return masses
    .map(mass => calculateTotalFuel(mass))
    .reduce((fuel, totalFuel) => totalFuel += fuel, 0);
}

const input = fs.readFileSync(path.resolve(__dirname, './input.txt'), 'utf8');
console.log('day1 part1:', part1(input));
console.log('day1 part2:', part2(input));

File: 2019/day1/day1.clj
'{:dependencies [[org.clojure/clojure "1.10.1"]]}
(require '[clojure.string :as str])

(def input
  (->>
   (slurp "2019/day1/input.txt")
   (str/split-lines)
   (map #(Integer/parseInt (str/trim %)))))

(defn calculate-fuel [mass] (- (quot mass 3) 2))

(defn calculate-total-fuel [mass]
  (let [fuel (calculate-fuel mass)]
    (if (< fuel 0)
      0
      (+ fuel (calculate-total-fuel fuel)))))

(defn part1 []
  (println "day1 part1:" (reduce + (map calculate-fuel input))))

(defn part2 []
  (println "day1 part2:" (reduce + (map calculate-total-fuel input))))

(part1)
(part2)

(comment
  "
   
  (defn calculate-total-fuel [mass]
    (reduce + (rest (take-while #(> % 0) (iterate calculate-fuel mass)))))
   
   #(> % 0) == pos? ;; Checks positive number

 ")
